fastFM/tests/test_ffm.py

```diff
@@ -38,5 +52,22 @@ def test_ffm2_predict_w0():
     y_pred = ffm2.ffm_predict(w0, w, V, X)
     assert_equal(y_pred, w0)
 
-if __name__ == '__main__':
-    test_ffm2_predict()
+def test_ffm2_fit():
+    w0, w, V, y, X = get_another_test_problem()
+    w0_init = w0
+    w_init = np.copy(w)
+    V_init = np.copy(V)
+    rank = 2
+
+    y_pred = ffm2.ffm_predict(w0, w, V, X)
+    msqr_before = mean_squared_error(y, y_pred)
+
+    w0, w, V = ffm2.ffm_als_fit(w0, w, V, X, y, rank)
+
+    y_pred = ffm2.ffm_predict(w0, w, V, X)
+    msqr_after = mean_squared_error(y, y_pred)
+
+    assert_true(w0 != w0_init)
+    assert_true(np.any(np.not_equal(w, w_init)))
```

Comment:
I think we could just use `assert  w != w_init`.
Same below

Reply:
assert(w != w_init)
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

Reply:
Okay, :-)  how about `assert np.all(w != w_init)`.

Reply:
wait. Should the every member of V array be changed anyway? (strange, but if i use 'all', this test fails.. Probably, i'm missing something.)

Reply:
If the values are randomly initialized I would indeed expect every value to be changed after one iteration except in pathological cases. 

Have a look at the values to see if something strange shows up. In the past this was often a hint that the memory mapping had issues (row major vs col major etc.).

 HTH

Reply:
Seems the mapping is fine, but only V[0,0] coef. changes without regularization.

Reply:
The tests are passing, so everything should be fine?
What about w0 and w? Can you add log output for data and parameter (before and after fit) on both the python and c++ site? 

I can't follow whats happening at the moment. This step is quite tricky and logging output helped me a lot in the past.

Reply:
I'll explain with logs..

For example:
```
PYTHON:
w0: 0
w: [ 0.  0.]
V: [[-2.4660091   0.03647256]
 [-1.11391683 -1.23835839]]

C++: 
w0: 0 
w: 0  0 
V:  -2.46601 0.0364726
 -1.11392  -1.23836
```
AFTER FIT
```
C++: 
w0: -92.4373 
w: 30.1384  49.6637 
V:   731.048 0.0364726
 -1.11392  -1.23836

PYTHON:
w0: -92.43728508040803
w: [ 30.13835662  49.66365965]
V: [[  7.31047901e+02   3.64725635e-02]
 [ -1.11391683e+00  -1.23835839e+00]]
```
As you can see, only V(0,0) have changed. And the test `assert np.all` for V will fail.

And if I hardcode on c++ side:
```   
settings->settings_.l2_reg_w1 = 0.1;
settings->settings_.l2_reg_w2 = 0.5;
```
We will have:

```
PYTHON:
w0: 0
w: [ 0.  0.]
V: [[ 0.12218913  1.21050484]
 [ 0.83090013  0.06547534]]
C++: 
w0: 0 
w: 0  0 
V:  0.122189    1.2105
   0.8309 0.0654753
```
AFTER FIT
```
C++: 
w0: -85.5123 
w: 28.4903 45.5889 
V:  23.2667  1.24985
 0.11857 0.154226

PYTHON:
w0: -85.51230850292907
w: [ 28.49029655  45.58891396]
V: [[ 23.2667138    1.24984746]
 [  0.11857022   0.15422599]]
```

Reply:
Thanks, now I can follow much better!

It's probably enough to use `settings->settings_.l2_reg_w2 = 0.5;`?
With regularization the parameters can change (due to shrinking) even when the parameters and data are not properly mapped (not sure if this is the case).

Its also possible that we have a bug on the c++ side. Can you try to reproduce the test settings with an additional test similar to https://github.com/ibayer/fastFM-core2/blob/master/fastFM_tests/api_tests/api_test.cpp#L52 . This would allow us to test the c++ side individually.

If the test passes we know that the mapping is at fault and otherwise I can debug the c++ site.

Reply:
```
api_test.cpp:151      0| 
BEFORE: W0: 0 
w: 0 0 
V:  0.680375 -0.211234
 0.566198   0.59688
```
AFTER FIT
```
AFTER: W0: -95.3092 
w: 30.7812
 51.113 
V:  -129.435 -0.211234
 0.566198   0.59688
```
So, the same in c++
Check branch https://github.com/ibayer/fastFM-core2/tree/test_addLog

Reply:
Great, that's a huge help. I'm very confident that the bug is not in the solver code.
I suspect mapping the raw memory into Eigen types causes the issue. The mapping is mostly done in https://github.com/ibayer/fastFM-core2/blob/master/fastFM/fastfm_impl.h, sadly not the clearest piece of code.

I'll take a closer took tonight.

Reply:
Hi, after a rather long debugging session I'm incline to belief that we don't have any bug at all.
What we observe is probably an artifact of our small toy example.

> The V parameter are optimized one at the time and it looks like the V[0,0] is able to capture all the residual so that nothing is left for the rest. Using regularization slows down the convergence (step size) and therefore other V's change the values to.

Let's just assume for now that his explanation is true.

To make it short, I think using regularization to ensure all V get changed is fine.
Sorry for the trouble. :-/

Reply:
"his explanation"? =)

Anyway, it's not a trouble. That's cool, if it's not a bug. 

As for the current branch, i see 2 preferred ways:
1. we change this toy example to something bigger, to get changes of each value.
2. we temporary leave the test as `any` with comment and fix them later, when we will have a simple way to change params from python side.

What i mean - it's strange to hardcode c++ values to get desired results for one special case in python, isn't it?

Reply:
Let's finish up this PR. I'm fine with 2, we can change the assert as soon as pass the settings to c++.